google/cloud/forseti/services/scanner/dao.py

```diff
@@ -18,150 +18,259 @@
+            .order_by(ScannerIndex.created_at_datetime.desc()).first())
+    return scanner_index.id if scanner_index else None
+
+
+class Violation(BASE):
+    """Row entry for a violation."""
+
+    __tablename__ = 'violations'
+
+    id = Column(Integer, primary_key=True)
+    created_at_datetime = Column(DateTime())
+    full_name = Column(String(1024))
+    resource_data = Column(Text(16777215))
+    resource_id = Column(String(256), nullable=False)
+    resource_type = Column(String(256), nullable=False)
+    rule_index = Column(Integer, default=0)
+    rule_name = Column(String(256))
+    scanner_index_id = Column(String(256))
+    violation_data = Column(Text)
+    violation_hash = Column(String(256))
+    violation_type = Column(String(256), nullable=False)
+
+    def __repr__(self):
+        """String representation.
+
+        Returns:
+            str: string representation of the Violation row entry.
+        """
+        string = ('<Violation(violation_type={}, resource_type={} '
+                  'rule_name={})>')
+        return string.format(
+            self.violation_type, self.resource_type, self.rule_name)
+
+class ViolationAccess(object):
+    """Facade for violations, implement APIs against violations table."""
+
+    def __init__(self, session):
+        """Constructor for the Violation Access.
+
+        Args:
+            session (Session): SQLAlchemy session object.
+        """
+        self.session = session
+
+    def create(self, violations, scanner_index_id):
+        """Save violations to the db table.
+
+        Args:
+            violations (list): A list of violations.
+            scanner_index_id (str): id of the `ScannerIndex` row for this
+                scanner run.
+        """
+        created_at_datetime = date_time.get_utc_now_datetime()
+        for violation in violations:
+            violation_hash = _create_violation_hash(
+                violation.get('full_name', ''),
+                violation.get('resource_data', ''),
+                violation.get('violation_data', ''),
+            )
+
+            violation = Violation(
+                created_at_datetime=created_at_datetime,
+                full_name=violation.get('full_name'),
+                resource_data=violation.get('resource_data'),
+                resource_id=violation.get('resource_id'),
+                resource_type=violation.get('resource_type'),
+                rule_index=violation.get('rule_index'),
+                rule_name=violation.get('rule_name'),
+                scanner_index_id=scanner_index_id,
+                violation_data=json.dumps(
+                    violation.get('violation_data')),
+                violation_hash=violation_hash,
+                violation_type=violation.get('violation_type')
+            )
+            self.session.add(violation)
+
+    def list(self, inv_index_id=None, scanner_index_id=None):
+        """List all violations from the db table.
+
```

Comment:
The documentation for this method deserves to be expanded, since there are some assumptions being made on how it's supposed to be used.  So, in the method doc, we should describe what is being returned, based on what the parameters are.  e.g. if `inventory_index_id` is given, then all the violations from all the scanner runs will be returned, even if `scanner_index_id` is given.  If you need to look for violations from a specific scanner index, then specify the parameter without inventory_index_id.  And if no parameter is given, than every violation will be returned. etc.

Reply:
Done.